src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -209,8 +179,9 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
             ConstructorInfo? constructorInfo = null;
             int?[]? parameterMap = null;
 
-            if (!TryFindPreferredConstructor(instanceType, argumentTypes, ref constructorInfo, ref parameterMap) &&
-                !TryFindMatchingConstructor(instanceType, argumentTypes, ref constructorInfo, ref parameterMap))
+            if (instanceType.IsAbstract ||
```

Comment:
Could you elaborate on the `IsAbstract` check? AFAIU, it was not preset in the `CreateFactory` flow. Was it implicitly covered by expression build failure?

Reply:
Yes, in case of `CreateFactory` and an abstract type it used to fail with the following exception:
```
System.InvalidOperationException: Can't compile a NewExpression with a constructor declared on an abstract class
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitNewExpression(Expression expr) in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\Compiler\LambdaCompiler.Expressions.cs:line 632
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags) in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\Compiler\LambdaCompiler.Generated.cs:line 120
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody(CompilerScope parent, Boolean inlined, CompilationFlags flags) in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\Compiler\LambdaCompiler.Lambda.cs:line 232
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody() in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\Compiler\LambdaCompiler.Lambda.cs:line 195
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda) in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\Compiler\LambdaCompiler.cs:line 187
   at System.Linq.Expressions.Expression`1.Compile() in C:\Git\runtime\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs:line 225
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateFactory(Type instanceType, Type[] argumentTypes) in C:\Git\runtime\src\libraries\Microsoft.Extensions.DependencyInjection.Abstractions\src\ActivatorUtilities.cs:line 104
   at Microsoft.Extensions.DependencyInjection.Specification.DependencyInjectionSpecificationTests.CreateInstanceFromFactory(IServiceProvider provider, Type type, Object[] args) in C:\Git\runtime\src\libraries\Microsoft.Extensions.DependencyInjection.Specification.Tests\src\ActivatorUtilitiesTests.cs:line 26
   at Microsoft.Extensions.DependencyInjection.Specification.DependencyInjectionSpecificationTests.<>c__DisplayClass29_0.<AAACreateInstance_WithAbstractTypeAndPublicConstructor_ThrowsCorrectException>b__0() in C:\Git\runtime\src\libraries\Microsoft.Extensions.DependencyInjection.Specification.Tests\src\ActivatorUtilitiesTests.cs:line 384
   at Xunit.Assert.RecordException(Func`1 testCode) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\Record.cs:line 50
```

Since abstract type is not desired in neither `CreateInstance` nor `CreateFactory` I assume it is fine to throw the same exception?

Reply:
I guess so, an explicit check beats an implicit check in my book.
It's up to maintainers to decide though, 'cause some software somewhere can handle the original exception with regards to the message.

Reply:
Makes sense. Let's see what others say.